lib/python/qmk/flashers.py

```diff
@@ -0,0 +1,126 @@
+import shutil
+import time
+
+import usb.core
+
+from qmk.constants import BOOTLOADER_VIDS_PIDS
+from milc import cli
+
+# yapf: disable
+_PID_TO_MCU = {
+    '2fef': 'atmega16u2',
+    '2ff0': 'atmega32u2',
+    '2ff3': 'atmega16u4',
+    '2ff4': 'atmega32u4',
+    '2ff9': 'at90usb64',
+    '2ffa': 'at90usb162',
+    '2ffb': 'at90usb128'
+}
+
+AVRDUDE_MCU = {
+    'atmega32a': 'm32',
+    'atmega328p': 'm328p',
+    'atmega328': 'm328',
+}
+# yapf: enable
+
+
+def _find_bootloader():
+    # To avoid running forever in the background, only look for bootloaders for 10min
+    start_time = time.time()
+    while time.time() - start_time < 600:
+        for bl in BOOTLOADER_VIDS_PIDS:
+            for vid, pid in BOOTLOADER_VIDS_PIDS[bl]:
+                vid_hex = int(f'0x{vid}', 0)
+                pid_hex = int(f'0x{pid}', 0)
+                dev = usb.core.find(idVendor=vid_hex, idProduct=pid_hex)
+                if dev:
+                    if bl == 'atmel-dfu':
+                        details = _PID_TO_MCU[pid]
+                    elif bl == 'caterina':
+                        details = (vid_hex, pid_hex)
+                    elif bl == 'hid-bootloader':
+                        if vid == '16c0' and pid == '0478':
+                            details = 'halfkay'
+                        else:
+                            details = 'qmk-hid'
+                    elif bl == 'stm32' or bl == 'apm32':
+                        details = (vid, pid)
+                    else:
+                        details = None
+                    return (bl, details)
+    return (None, None)
+
+
+def _find_serial_port(vid, pid):
+    if 'windows' in cli.platform.lower():
+        from serial.tools.list_ports_windows import comports
+    else:
+        from serial.tools.list_ports_posix import comports
+    for port in comports():
+        port, desc, hwid = port
+        if f'{vid:04x}:{pid:04x}' in hwid.casefold():
+            return port
+    return None
+
+
+def _flash_caterina(details, file):
+    port = _find_serial_port(details[0], details[1])
```

Comment:
Maybe this needs a waiting loop in case the USB device has been detected, but the serial port has not been initialized yet? And some clean error exit if the port detection fails (instead of just trying to run a broken command).

The code in `platforms/avr/flash.mk` also waits until the port becomes writable (this might be needed because of some races with udev; it will also fail if udev rules are missing).

Reply:
Good points, I'll see what I can do.

> it will also fail if udev rules are missing

I think everything else will also fail in that case. :D

Reply:
On Posix systems now the code checks if the port is writable. On Windows it waits 1 second. (similarly to flash.mk)

```
$ qmk flash lets_split_rev2_default.hex
Flashing binary firmware...
Please reset your keyboard into bootloader mode now!
Press Ctrl-C to exit.

The Caterina bootloader was found but is not writable. Check 'qmk doctor' output for advice.
```